// prettier-plugin-solidity/dist/slang-nodes/BitwiseAndExpression.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { printBinaryOperation } from '../slang-printers/print-binary-operation.js';
import { createHugFunction } from '../slang-utils/create-hug-function.js';
import { createKindCheckFunction } from '../slang-utils/create-kind-check-function.js';
import { extractVariant } from '../slang-utils/extract-variant.js';
import { SlangNode } from './SlangNode.js';
import { Expression } from './Expression.js';
const tryToHug = createHugFunction(['+', '-', '*', '/', '**', '<<', '>>']);
const printBitwiseAndExpression = printBinaryOperation(createKindCheckFunction([
    NonterminalKind.InequalityExpression,
    NonterminalKind.EqualityExpression,
    NonterminalKind.AndExpression,
    NonterminalKind.OrExpression
]));
export class BitwiseAndExpression extends SlangNode {
    constructor(ast, options) {
        super(ast);
        this.kind = NonterminalKind.BitwiseAndExpression;
        this.leftOperand = extractVariant(new Expression(ast.leftOperand, options));
        this.operator = ast.operator.unparse();
        this.rightOperand = extractVariant(new Expression(ast.rightOperand, options));
        this.updateMetadata(this.leftOperand, this.rightOperand);
        this.leftOperand = tryToHug(this.leftOperand);
        this.rightOperand = tryToHug(this.rightOperand);
    }
    print(path, print, options) {
        return printBitwiseAndExpression(this, path, print, options);
    }
}
