// prettier-plugin-solidity/dist/slang-nodes/EqualityExpression.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { createHugFunction } from '../slang-utils/create-hug-function.js';
import { createKindCheckFunction } from '../slang-utils/create-kind-check-function.js';
import { printBinaryOperation } from '../slang-printers/print-binary-operation.js';
import { extractVariant } from '../slang-utils/extract-variant.js';
import { SlangNode } from './SlangNode.js';
import { Expression } from './Expression.js';
const tryToHug = createHugFunction(['==', '!=']);
const printEqualityExpression = printBinaryOperation(createKindCheckFunction([
    NonterminalKind.AndExpression,
    NonterminalKind.OrExpression
]));
export class EqualityExpression extends SlangNode {
    constructor(ast, options) {
        super(ast);
        this.kind = NonterminalKind.EqualityExpression;
        this.leftOperand = extractVariant(new Expression(ast.leftOperand, options));
        this.operator = ast.operator.unparse();
        this.rightOperand = extractVariant(new Expression(ast.rightOperand, options));
        this.updateMetadata(this.leftOperand, this.rightOperand);
        this.leftOperand = tryToHug(this.leftOperand);
    }
    print(path, print, options) {
        return printEqualityExpression(this, path, print, options);
    }
}
