// prettier-plugin-solidity/dist/slang-nodes/ExponentiationExpression.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { doc } from 'prettier';
import { createBinaryOperationPrinter } from '../slang-printers/create-binary-operation-printer.js';
import { binaryIndentRulesBuilder } from '../slang-printers/print-binary-operation.js';
import { createHugFunction } from '../slang-utils/create-hug-function.js';
import { createKindCheckFunction } from '../slang-utils/create-kind-check-function.js';
import { extractVariant } from '../slang-utils/extract-variant.js';
import { SlangNode } from './SlangNode.js';
import { Expression } from './Expression.js';
const { group } = doc.builders;
const tryToHug = createHugFunction(['**']);
const shouldIndent = createKindCheckFunction([
    NonterminalKind.MultiplicativeExpression,
    NonterminalKind.AdditiveExpression,
    NonterminalKind.ShiftExpression,
    NonterminalKind.BitwiseAndExpression,
    NonterminalKind.BitwiseOrExpression,
    NonterminalKind.BitwiseXorExpression,
    NonterminalKind.InequalityExpression,
    NonterminalKind.EqualityExpression,
    NonterminalKind.AndExpression,
    NonterminalKind.OrExpression
]);
const printExponentiationExpression = createBinaryOperationPrinter(() => (document) => group(document), // always group
binaryIndentRulesBuilder(shouldIndent) // indent as a binary operation with some exceptions
);
export class ExponentiationExpression extends SlangNode {
    constructor(ast, options) {
        super(ast);
        this.kind = NonterminalKind.ExponentiationExpression;
        this.leftOperand = extractVariant(new Expression(ast.leftOperand, options));
        this.operator = ast.operator.unparse();
        this.rightOperand = extractVariant(new Expression(ast.rightOperand, options));
        this.updateMetadata(this.leftOperand, this.rightOperand);
        this.rightOperand = tryToHug(this.rightOperand);
        this.leftOperand = tryToHug(this.leftOperand);
    }
    print(path, print, options) {
        return printExponentiationExpression(this, path, print, options);
    }
}
