// prettier-plugin-solidity/dist/slang-nodes/OrExpression.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { printLogicalOperation } from '../slang-printers/print-logical-operation.js';
import { createHugFunction } from '../slang-utils/create-hug-function.js';
import { extractVariant } from '../slang-utils/extract-variant.js';
import { SlangNode } from './SlangNode.js';
import { Expression } from './Expression.js';
const tryToHug = createHugFunction(['&&']);
export class OrExpression extends SlangNode {
    constructor(ast, options) {
        super(ast);
        this.kind = NonterminalKind.OrExpression;
        this.leftOperand = extractVariant(new Expression(ast.leftOperand, options));
        this.operator = ast.operator.unparse();
        this.rightOperand = extractVariant(new Expression(ast.rightOperand, options));
        this.updateMetadata(this.leftOperand, this.rightOperand);
        this.leftOperand = tryToHug(this.leftOperand);
        this.rightOperand = tryToHug(this.rightOperand);
    }
    print(path, print, options) {
        return printLogicalOperation(this, path, print, options);
    }
}
