// prettier-plugin-solidity/dist/slang-nodes/Parameter.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { doc } from 'prettier';
import { joinExisting } from '../slang-utils/join-existing.js';
import { extractVariant } from '../slang-utils/extract-variant.js';
import { SlangNode } from './SlangNode.js';
import { TypeName } from './TypeName.js';
import { StorageLocation } from './StorageLocation.js';
import { TerminalNode } from './TerminalNode.js';
const { group } = doc.builders;
export class Parameter extends SlangNode {
    constructor(ast, options) {
        super(ast);
        this.kind = NonterminalKind.Parameter;
        this.typeName = extractVariant(new TypeName(ast.typeName, options));
        if (ast.storageLocation) {
            this.storageLocation = new StorageLocation(ast.storageLocation);
        }
        if (ast.name) {
            this.name = new TerminalNode(ast.name);
        }
        this.updateMetadata(this.typeName, this.storageLocation);
    }
    print(path, print) {
        return group(joinExisting(' ', [
            path.call(print, 'typeName'),
            path.call(print, 'storageLocation'),
            path.call(print, 'name')
        ]));
    }
}
