// prettier-plugin-solidity/dist/slang-nodes/ShiftExpression.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { printBinaryOperation } from '../slang-printers/print-binary-operation.js';
import { createHugFunction } from '../slang-utils/create-hug-function.js';
import { createKindCheckFunction } from '../slang-utils/create-kind-check-function.js';
import { extractVariant } from '../slang-utils/extract-variant.js';
import { SlangNode } from './SlangNode.js';
import { Expression } from './Expression.js';
const tryToHugLeftOperand = createHugFunction([
    '+',
    '-',
    '*',
    '/',
    '**',
    '<<',
    '>>'
]);
const tryToHugRightOperand = createHugFunction(['+', '-', '*', '/', '**']);
const printShiftExpression = printBinaryOperation(createKindCheckFunction([
    NonterminalKind.BitwiseAndExpression,
    NonterminalKind.BitwiseOrExpression,
    NonterminalKind.BitwiseXorExpression,
    NonterminalKind.InequalityExpression,
    NonterminalKind.EqualityExpression,
    NonterminalKind.AndExpression,
    NonterminalKind.OrExpression
]));
export class ShiftExpression extends SlangNode {
    constructor(ast, options) {
        super(ast);
        this.kind = NonterminalKind.ShiftExpression;
        this.leftOperand = extractVariant(new Expression(ast.leftOperand, options));
        this.operator = ast.operator.unparse();
        this.rightOperand = extractVariant(new Expression(ast.rightOperand, options));
        this.updateMetadata(this.leftOperand, this.rightOperand);
        this.leftOperand = tryToHugLeftOperand(this.leftOperand);
        this.rightOperand = tryToHugRightOperand(this.rightOperand);
    }
    print(path, print, options) {
        return printShiftExpression(this, path, print, options);
    }
}
