// prettier-plugin-solidity/dist/slang-nodes/TupleDeconstructionStatement.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { printGroupAndIndentIfBreakPair } from '../slang-printers/print-group-and-indent-if-break-pair.js';
import { extractVariant } from '../slang-utils/extract-variant.js';
import { SlangNode } from './SlangNode.js';
import { TupleDeconstructionElements } from './TupleDeconstructionElements.js';
import { Expression } from './Expression.js';
export class TupleDeconstructionStatement extends SlangNode {
    constructor(ast, options) {
        var _a;
        super(ast);
        this.kind = NonterminalKind.TupleDeconstructionStatement;
        this.varKeyword = (_a = ast.varKeyword) === null || _a === void 0 ? void 0 : _a.unparse();
        this.elements = new TupleDeconstructionElements(ast.elements, options);
        this.expression = extractVariant(new Expression(ast.expression, options));
        this.updateMetadata(this.elements, this.expression);
    }
    print(path, print) {
        return printGroupAndIndentIfBreakPair([this.varKeyword ? 'var (' : '(', path.call(print, 'elements'), ') = '], [path.call(print, 'expression'), ';']);
    }
}
