// prettier-plugin-solidity/dist/slang-nodes/YulForStatement.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { doc } from 'prettier';
import { extractVariant } from '../slang-utils/extract-variant.js';
import { SlangNode } from './SlangNode.js';
import { YulBlock } from './YulBlock.js';
import { YulExpression } from './YulExpression.js';
const { join } = doc.builders;
export class YulForStatement extends SlangNode {
    constructor(ast, options) {
        super(ast);
        this.kind = NonterminalKind.YulForStatement;
        this.initialization = new YulBlock(ast.initialization, options);
        this.condition = extractVariant(new YulExpression(ast.condition, options));
        this.iterator = new YulBlock(ast.iterator, options);
        this.body = new YulBlock(ast.body, options);
        this.updateMetadata(this.initialization, this.condition, this.iterator, this.body);
    }
    print(path, print) {
        return join(' ', [
            'for',
            path.call(print, 'initialization'),
            path.call(print, 'condition'),
            path.call(print, 'iterator'),
            path.call(print, 'body')
        ]);
    }
}
